### DISCUSSION TITLE
Allow `unsafe` code in iterators

### DISCUSSION THREAD

**Discussion by ero-qt** [dotnet/csharplang#6265] [Language Ideas]
I've been wondering for forever why `unsafe` code is disallowed in iterators. It's a massive hassle having to create my own class that implements `IEnumerator<T>, IEnumerator, IEnumerable<T>, IEnumerable` every single time I would like to use unsafe code in a method that's meant to yield its results.

Example:
```cs
unsafe IEnumerable<T> Iter<T>(byte* data, int count) where T : unmanaged
{
  for (int i = 0; i < count; i++)
  {
    yield return *(T*)(data + i * sizeof(T));
  }
}
```
This will not compile with the message `Unsafe code may not appear in iterators`. The current work-around is a class like the following:

```cs
unsafe class Iter<T> : IEnumerator<T>, IEnumerator, IEnumerable<T>, IEnumerable
  where T : unmanaged
{
  private readonly byte[] _data;
  private readonly int _count;

  private T _currentItem;
  private int _currentIndex;

  public Iter(byte[] data, int count)
  {
    _data = data;
    _count = count;
  }

  public object Current => _currentItem;
  T IEnumerator<T>.Current => _currentItem;

  public bool MoveNext()
  {
    if (_currentIndex >= _count)
    {
      return false;
    }

    fixed (byte* pData = _data)
    {
      _currentItem = *(T*)(pData + _currentIndex * sizeof(T));
    }

    _currentIndex++;

    return true;
  }

  public void Reset()
  {
    _currentItem = default;
    _currentIndex = 0;
  }

  public IEnumerator GetEnumerator() => this;
  IEnumerator<T> IEnumerable<T>.GetEnumerator() => this;

  public void Dispose() { }
}
```
With an implementation like the following:
```cs
IEnumerable<T> Iter<T>(byte[] data, int count) where T : unmanaged
{
  foreach (T t in new Iter<T>(data, count))
  {
    yield return t;
  }
}
```
---
This is quite clearly a huge amount of boilerplate and overhead. I'm wondering why this has not been addressed?

**Comment by ero-qt** [comment]
Disregard the fact that I pass a `byte[]` in the second version, I couldn't think of a better example on the spot. Obviously there's better ways to get an array of type `T` from an array of raw bytes. Though this does go to show that `unsafe` iterators are a must-have; if you do happen to work with pointers, you cannot pass those pointers into the constructor of the `Iter<T>` class, since that is also considered `unsafe`.

**Comment by Xyncgas** [comment]
Well maybe because there is compiler magic that makes use of call backs the moment you use the yield keyword, and they don't want you to use pointers because when your code resumes via callback (by yield executes) your pointer might not point to the same thing

**Reply by alexrp** [reply]
I think if you're in `unsafe` land, the assumption should be that you know what you're doing.

Maybe there are corner cases I haven't thought of, but I think the only big concerns are:

1. Taking the address of a local and using it after a `yield` point.
2. `fixed` statements that contain `yield` statements.

The language could be updated to disallow these specifically rather than making all unsafe code illegal in iterators.

**Reply by Xyncgas** [reply]
Apparently, I like playing with pointer, but isn't it time to respect the fact that we have an array pool and a modern language...
Apparently, if you have an array pool your pointers are all gonna dangle after garbage collection where the objects are MOVED in memory for performance reasons.
Apparently C# fixes this for you by repointing all the reference you had for the objects it's moving to the new location, but if you had a pointer in unsafe code, they are not going to fix that for you.
I think ** Taking the address of a local and using it after a yield point. ** is not gonna work because you meant taking the address of your local in the memory right well that possibly changes once you yield, and depends on how corotines are implemented it may or may not have a stack meaning all your local variable could be actually just variables of an instance created automatically, or it may has a stack preserved in the coroutine implementation, in that case you can use pointers that points to a place in stack where your locals are.

**Comment by siegfriedpammer** [comment]
You can read some of the reasons on Eric Lippert's blog: https://docs.microsoft.com/en-us/archive/blogs/ericlippert/iterator-blocks-part-six-why-no-unsafe-code

**Reply by alexrp** [reply]
> First, it is an incredibly unlikely scenario.

This might have been true in 2009, but it is certainly not true today. The language and surrounding ecosystem have gotten much friendlier towards unsafe code. I have multiple projects where I have to resort to silly workarounds because I can't use pointers in iterators.

> Second, the scenario is a violent mixing of "levels."

I similarly find this argument unconvincing today. C# has very much embraced the fact that its users want to be able to write high-performance unsafe code.

> Third, we would have to do some extra work to make "fixed" blocks work correctly, work that would be directly opposed to our guidelines for the correct and efficient use of the "fixed" block.

This, however, I find to be a very valid concern. I know the Roslyn developers don't have much appetite for touching the `fixed` code in the compiler. But I have run into exactly zero cases where I wanted to use `fixed` in an iterator. I think, more often than not, if someone wants to use unsafe code in an iterator, it's because they're dealing with pointers to unmanaged memory. So I think it would be perfectly reasonable to just specifically disallow `fixed` statements in iterators.

**Reply by ero-qt** [reply]
Is this third point really a valid concern across the board, though? Is there a problem with the following approach?
```cs
IEnumerable<T> Iter<T>(byte[] data, int count) where T : unmanaged
{
  fixed (byte* pData = data)
  {
    for (int i = 0; i < count; i++)
    {
      yield return *(T*)(pData + i * sizeof(T));
    }
  }
}
```

**Reply by svick** [reply]
@alexrp 

> C# has very much embraced the fact that its users want to be able to write high-performance unsafe code.

C# has indeed embraced the fact that users want to write high-performance code, but tries to ensure it's still safe where possible. That's why you'll see a lot of `Span<T>` or `Memory<T>` in modern C# code, but not a lot of pointers.

Also, if you want to write high-performance code, you should probably avoid `IEnumerable<T>`, since it involves a decent amount of overhead.

@just-ero 

> Is there a problem with the following approach?

The problem is exactly what Eric Lippert pointed out in his blog post: It would require using a heavy-weight, long-lasting `GCHandle` object, instead of the light-weight, short-term `pinned` local, which is used by `fixed` statements today.

**Reply by ero-qt** [reply]
A code that makes use of `Span<T>` or `Memory<T>` may not require `AllowUnsafeBlocks`, that might be true. But on some level its implementations all use unsafe code, no? (see: [@tannergooding on Twitter](https://twitter.com/tannergooding/status/1541531327307976709))

> Also, if you want to write high-performance code, you should probably avoid `IEnumerable<T>`, since it involves a decent amount of overhead.

Writing any iterator creates almost exactly the class I described in my original post. I'm not sure how the overhead possibly matters? (see: [sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRhQG4BYAKHIIGYAeASwDsAXAPhwEkmYoAKASnIBvcjlE4AZtBw9GTHHRwBeHFiLycNHPlXyA1LoFkxxvPnx4A7PNJkAvkA))

**Reply by svick** [reply]
> A lot (if not most) code that make use of `Span<T>` or `Memory<T>` don't require `AllowUnsafeBlocks`, that may be true. But on some level its implementations all use unsafe code, no?

That depends. If you use `Span<T>` to refer to a part of an array, that's not unsafe. If you `stackalloc` a `Span<T>`, that's not unsafe. If you turn a pointer into a `Span<T>`, that's unsafe, but you're limiting it to a single, easy to verify unsafe operation. If you use `Unsafe`, then that's indeed unsafe (hence the name), but I think the vast majority of uses of `Span<T>` don't need that (or again, limit its uses as much as possible).

> Writing any iterator creates almost exactly the class I described in my original post. I'm not sure how the overhead possibly matters? 

I'm saying that if you're writing high-performance code, `IEnumerable<T>` is probably the wrong abstraction to use. And both the hypothetical `unsafe` iterator method and your `class Iter<T>` are equally bad.

**Reply by alexrp** [reply]
@svick

> C# has indeed embraced the fact that users want to write high-performance code, but tries to ensure it's still safe where possible.

And that can be done here too. For example, it could simply be illegal to take the address of a local in an iterator.

> That's why you'll see a lot of `Span<T>` or `Memory<T>` in modern C# code, but not a lot of pointers.

This is true only for the most trivial cases of passing buffers around. The language has made it easier with each new version to work with pointers and blittable data: `unmanaged` constraint, function pointers, `nint`/`nuint`, pattern-based `fixed`, etc. If you look to some modern interop libraries/tools such as [TerraFX](https://github.com/terrafx) and [CsWin32](https://github.com/microsoft/CsWin32), you'll note that working with pointers to blittable structures is increasingly becoming the norm. It should also be noted that `Span<T>`/`Memory<T>` are inadequate for some scenarios because of their `int Length`.

> Also, if you want to write high-performance code, you should probably avoid `IEnumerable<T>`, since it involves a decent amount of overhead.

Depends on the scenario. For example, if I'm going over the list of processes, threads, modules, and heaps in a snapshot of the system, I don't want to retrieve information for all of them, just the few I care about. That's a good use case for an iterator.

> If you `stackalloc` a Span<T>, that's not unsafe.

Note: Only as long as `[SkipLocalsInit]` is not in play somewhere that affects your code. It's probably best to always assume that `stackalloc`'d memory is undefined, even with `Span<T>`.

**Comment by CyrusNajmabadi** [comment]
>  I'm wondering why this has not been addressed?

The complexity is high, and the gain is low. As an example, you're the first person to ask for this in the history of Roslyn. 

If there is high community interest though, and we can like we could solve the complexity issues, it's something we might be able to do.

**Reply by Xyncgas** [reply]
When are we gonna get readonly locals

**Reply by CyrusNajmabadi** [reply]
@Xyncgas do not spam unrelated issues.

**Comment by tfenise** [comment]
I think as a workaround, one could use helper methods. For example, instead of
```C#
unsafe IEnumerable<T> Iter<T>(byte* data, int count) where T : unmanaged
{
  for (int i = 0; i < count; i++)
  {
    yield return *(T*)(data + i * sizeof(T));
  }
}
```
write:
```C#
unsafe T Read<T>(IntPtr data, int i) where T : unmanaged => *(T*)((byte*)data + i * sizeof(T));

unsafe IEnumerable<T> Iter<T>(byte* data, int count) where T : unmanaged => Iter_Helper<T>((IntPtr)data, count);

IEnumerable<T> Iter_Helper<T>(IntPtr data, int count) where T : unmanaged
{
  for (int i = 0; i < count; i++)
  {
    yield return Read<T>(data, i);
  }
}
```

It is not as ideal as one may expect, but it is simpler than writing an implementation of `IEnumerator<T>` by hand.

**Reply by ero-qt** [reply]
[Benjamin Moir](https://github.com/DaZombieKiller) suggested this code on the [C# Discord server](https://discord.gg/csharp):
```cs
unsafe IEnumerable<T> Iter<T>(byte* data, int count) where T : unmanaged
{
  return IterCore<T>((nint)data, count);
}

IEnumerable<T> IterCore<T>(nint ptr, int count) where T : unmanaged
{
  for (int i = 0; i < count; i++)
  {
    yield return Unsafe.Add(ref AsRef<T>(ptr), i);
  }
}

static unsafe ref T AsRef<T>(nint ptr)
{
  return ref Unsafe.AsRef<T>((void*)ptr);
}
```
I'm not sure if this is any better or worse than what you suggested.

**Reply by CyrusNajmabadi** [reply]
I'm confused why an iterator it's being used here instead of just using Span/Memory to represent this sequence of data

**Reply by PathogenDavid** [reply]
I agree with Cyrus, your specific examples so far are just inefficient iterating variants of `MemoryMarshal.Cast`.

**Reply by ero-qt** [reply]
I apologize for the poor choice of examples. The discussion is about allowing `unsafe` code in iterators, not about solving my examples in different ways.

**Comment by MillKaDe** [comment]
As far as I understand, you want to reinterpret a bunch of bytes as a bunch of `T`s ?
`ReadOnlySpan` already supports enumeration via `foreach` [SharpLab](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUBgZgAIBvPIo1TANiOgGcBDAMwFNKAWIgdTAEtgrAIL0APABUAfEQAUqDgCoifAA4xlUYETh8A5gPrq+molAgBbAJREA7gAtWYduKIg6Uc4yiNdrOKXIKIgA3RjAiemYiAF4iABIAIhISAAYAXzS3FIAPZMwQAA0SHX1gegy0hIBuQIoqAE5ZRMZ6UjNzNKJs0mAATxVWAHsomXFLADoAOUZzVk6ZEno+AC8hkbHOgCNewXpLdWZBsE9QIgByReY0s4TLGtwgjS0+OG7YlPvH0PC+ehUY0ysGxEABKrEYcAA8lAADa9ADKKi8EmkMlU6nad1qREOTkYAGM7LJvkRnlBlH9rA1ZJEjK9EIh1MAsQ8iGlAqhiAwWOwqLR5EQALKMYyyaxkVl1RTKf6xejAAkAa0YMJhg3xRG2ggA2gBdUiYdTodQEdQcdQAVnU1HUAHZ1AAONKfIK8ATCMQQYzAVHoogOo13CjYt2CESiCD0OxHH2yP3mogcINEEP8MNiLWsX1qIjG/3JwLs3BpIA=) :
```c#
using System;
public static class Program {
  static unsafe void WriteAs<T> (void* ip, int digits, int num) where T : unmanaged {
    var sf = $"{{0}}: 0x{{1:X{digits}}}";
    Console.WriteLine ($"as {num} x {typeof (T).Name} ({sizeof (T)} bytes), format: '{sf}'");
    int idx = 0;
    var isp = new ReadOnlySpan<T> (ip, num);
    foreach (var t in isp) Console.WriteLine (sf, idx++, t);
  }
  public unsafe static void Main () {
    void* ip = stackalloc byte[] {1, 2, 3, 4, 5, 6, 7, 8};
    WriteAs<uint> (ip, 8, 2);  
    WriteAs<ushort> (ip, 4, 4);  
    WriteAs<byte> (ip, 2, 8);  
  }
}
```
.. which prints :
```
as 2 x UInt32 (4 bytes), format: '{0}: 0x{1:X8}'
0: 0x04030201
1: 0x08070605

as 4 x UInt16 (2 bytes), format: '{0}: 0x{1:X4}'
0: 0x0201
1: 0x0403
2: 0x0605
3: 0x0807

as 8 x Byte (1 bytes), format: '{0}: 0x{1:X2}'
0: 0x01
1: 0x02
2: 0x03
3: 0x04
4: 0x05
5: 0x06
6: 0x07
7: 0x08
```

I admit that doesn't answer your 'unsafe iterator' question, but maybe it solves the original problem you tried to solve by writing an iterator.

**Reply by ero-qt** [reply]
No. This was poorly-chosen sample code. The discussion is about allowing `unsafe` code in iterators in general, not this specific example.

**Reply by CyrusNajmabadi** [reply]
What is an example that demonstrates the need for this?  Thanks!

**Comment by Xyncgas** [comment]
It's very painful, writing codes managing memory instead of focusing on productivity, I thought we made the cpu we built the operating system we got a manage array pool so we can go ahead playing video games running mass surveillance.
C++ is fun, before you have to write a library yourself trying to do everything, for one particular project like building a website that runs on webassembly via Emscripten that has dynamic content, then you start looking for a new place called home, where the compiler does more for you so you can use a easier language.
I am thankful I am not suffering with c++
